[JSC] Materialize Const128 as MemoryValue

Constellation opened the pull request:
#### 5ec607384e6cab3e0822efcc9f6bc3227b7d5505
<pre>
[JSC] Materialize Const128 as MemoryValue
<a href="https://bugs.webkit.org/show_bug.cgi?id=249372">https://bugs.webkit.org/show_bug.cgi?id=249372</a>
rdar://103388495

Reviewed by Justin Michaud.

This patch implements Const128 MemroyValue materialization in B3 MoveConstants phase.
Materializing Const128 takes a lot of instructions. So usually compiler makes it
load-from-data-section. We already did this for double / float, so this patch extends
B3 MoveConstants to handle Const128 too.

We reorganize our DataSection in B3. Previously all the slots are 8-bytes. But now,
we create this section as follows.

    [ V128 slots ... ][ Double slots ... ][ Float slots ... ]

Since the section starts with V128, which has the largest alignment, we do not need to
consider about alignment when calculating offset in this table. Each slot is exact-sized
so that we do not waste much memory for double / float even though there is a V128 slot.

* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3MoveConstants.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::dump const):
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
(JSC::B3::ValueKey::materialize const):
* Source/JavaScriptCore/b3/B3ValueKey.h:
(JSC::B3::ValueKey::ValueKey):
(JSC::B3::ValueKey::vectorValue const):
(JSC::B3::ValueKey::hash const):
* Source/JavaScriptCore/jit/SIMDInfo.h:
(JSC::bitEquals):

Canonical link: <a href="https://commits.webkit.org/257939@main">https://commits.webkit.org/257939@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2026d08d288e900072f0e7dd3ec35d9f1ac1d33f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100384 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9546 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33447 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109700 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169939 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10458 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/100 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92788 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107578 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106162 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7915 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91183 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34570 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22573 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90927 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3287 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24092 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/86943 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/742 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3279 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/217 "Found 1 new test failure: http/tests/security/contentSecurityPolicy/source-list-parsing-05.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29115 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9400 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43582 "Passed tests") | [❌ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/89826 "Hash 2026d08d for PR 7671 does not build (failure)") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5095 "Built successfully") | | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/37/builds/89826 "Hash 2026d08d for PR 7671 does not build (failure)") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2835 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -4075,14 +4075,8 @@ class LowerToAir {
-            append(Air::VectorReplaceLaneInt64, Arg::imm(0), a, result);
-            append(Move, Arg::bigImm(value.u64x2[1]), a);
-            append(Air::VectorReplaceLaneInt64, Arg::imm(1), a, result);
+            // We expect that the moveConstants() phase has run, and any constant vector referenced from stackmaps get fused.
```
Should we assert not reached?

Constellation replied:
We are not moving Const128 for zero case (as the same to ConstFloat / ConstDouble) since vxorps is faster and compact. So we need this `MoveZeroToVector`.

justinmichaud reviewed (approved):
r=me

Constellation committed:
[JSC] Materialize Const128 as MemoryValue
https://bugs.webkit.org/show_bug.cgi?id=249372
rdar://103388495

Reviewed by Justin Michaud.

This patch implements Const128 MemroyValue materialization in B3 MoveConstants phase.
Materializing Const128 takes a lot of instructions. So usually compiler makes it
load-from-data-section. We already did this for double / float, so this patch extends
B3 MoveConstants to handle Const128 too.

We reorganize our DataSection in B3. Previously all the slots are 8-bytes. But now,
we create this section as follows.

    [ V128 slots ... ][ Double slots ... ][ Float slots ... ]

Since the section starts with V128, which has the largest alignment, we do not need to
consider about alignment when calculating offset in this table. Each slot is exact-sized
so that we do not waste much memory for double / float even though there is a V128 slot.

* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3MoveConstants.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::dump const):
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
(JSC::B3::ValueKey::materialize const):
* Source/JavaScriptCore/b3/B3ValueKey.h:
(JSC::B3::ValueKey::ValueKey):
(JSC::B3::ValueKey::vectorValue const):
(JSC::B3::ValueKey::hash const):
* Source/JavaScriptCore/jit/SIMDInfo.h:
(JSC::bitEquals):

Canonical link: https://commits.webkit.org/257939@main
